backend/cmd/headlamp.go

```diff
@@ -605,9 +611,168 @@ func createHeadlampHandler(config *HeadlampConfig) http.Handler {
 	return r
 }
 
+func parseClusterAndToken(r *http.Request) (string, string) {
+	cluster := ""
+	re := regexp.MustCompile(`^/clusters/([^/]+)/.*`)
+	urlString := r.URL.RequestURI()
+
+	matches := re.FindStringSubmatch(urlString)
+	if len(matches) > 1 {
+		cluster = matches[1]
+	}
+
+	// get token
+	token := r.Header.Get("Authorization")
+	token = strings.TrimPrefix(token, "Bearer ")
+
+	return cluster, token
+}
+
+func isTokenAboutToExpire(token string) bool {
+	const TokenParts = 3
+
+	// parse expiry time from token
+	parts := strings.Split(token, ".")
+	if len(parts) != TokenParts {
+		return false
+	}
+
+	payloadPart := parts[1]
+
+	payloadBytes, err := base64.RawStdEncoding.DecodeString(payloadPart)
+	if err != nil {
+		return false
+	}
+
+	var payload map[string]interface{}
+	if err := json.Unmarshal(payloadBytes, &payload); err != nil {
+		return false
+	}
+
+	// check if token is expired
+	exp, ok := payload["exp"].(float64)
+	if !ok {
+		return false
+	}
+
+	// if token is not about to expire, then skip
+	expTime := time.Unix(int64(exp), 0)
+
+	return time.Until(expTime) <= time.Second*10
```

Comment:
Any particular reason why we're using 10s as the check? Not sure how long the tokens are valid for on average but we probably should try to renew it before such a short period.